### ISSUE TITLE
Implement recommended best practises to prevent abuse rate limits

### ISSUE THREAD

**Issue by bkeepers** [octokit/octokit.js#526]
It would be great if this library had an easy way to handle each of the recommendations from GitHub's [best practices for integrators](https://developer.github.com/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits) on dealing with abuse rate limits. I wanted to open an issue to talk through some ideas to see if you would be open to some PRs that start to move toward this.

The docs say:

> [Abuse rate limits](https://developer.github.com/v3/#abuse-rate-limits) are another way we ensure the API's availability. To avoid hitting this limit, you should ensure your application follows the guidelines below.
> 
> -   Make authenticated requests, or use your application's client ID and secret. Unauthenticated requests are subject to more aggressive abuse rate limiting.

✅ Obviously already support for this.

> -   Make requests for a single user or client ID serially. Do not make requests for a single user or client ID concurrently.

Ideally the constructor would take a `queue` argument, which could either be `true` to use an internal queue, or an object that responds to `push`, such as [async.queue](http://caolan.github.io/async/docs.html#queue).

```js
var GitHubApi = require("github");

var github = new GitHubApi({
    queue: true
});
```

> -   If you're making a large number of `POST`, `PATCH`, `PUT`, or `DELETE` requests for a single user or client ID, wait at least one second between each request.

If the previous queue proposal is used, it would be easy for either the internal queue implementation or an external one to set timeouts between requests.

> -   When you have been limited, use the `Retry-After` response header to slow down. The value of the `Retry-After` header will always be an integer, representing the number of seconds you should wait before making requests again. For example, `Retry-After: 30` means you should wait 30 seconds before sending more requests.

This starts to get a little tricky. It's probably best left up to the caller to handle this, but it might be worth exploring some ways to make it easier to handle.

> -   Requests that create content which triggers notifications, such as issues, comments and pull requests, may be further limited and will not include a `Retry-After` header in the response. Please create this content at a reasonable pace to avoid further limiting.

This is also solved by queue proposal.

Also missing from the best practices guide is recommendations on ETags. It would be great to see some sort of ETag handling like #255.

What do you think? I am up for trying to implement some of these recommendations, but wanted to chat through them before starting to open PRs.

**Comment by jeffwilcox** [comment]
We're successfully using `node-github` with ETags and all the other GitHub recommendations, but have chosen to implement this a layer above - namely, we need to maintain our own cache (Redis typically) of long-lived, best-known GitHub requests; this helps land a solid 304 story.

Happy to try to share what we've done to help here, Brandon... but not sure how it fits into the rather low-level approach of the library just wrapping request.

**Comment by bkeepers** [comment]
@jeffwilcox Hey Jeff, I'd love to see what you've done!

Here's what I just did in https://github.com/probot/probot/pull/109 to add throttling:

```js
const Bottleneck = require('bottleneck');

function rateLimitedClient(github) {
  const limiter = new Bottleneck(1, 1000);
  const oldHandler = github.handler;
  github.handler = (msg, block, callback) => {
    limiter.submit(oldHandler.bind(github), msg, block, callback);
  };
  return github;
}
```

**Comment by jeffmcaffer** [comment]
You may also want to look at https://github.com/Microsoft/ghrequestor.  This is the underpinning of https://github.com/Microsoft/ghcrawler. While it is only designed for GETs, some of the external caching, etag handling and multipage support may be relevant.

**Comment by jeffwilcox** [comment]
Too many Jeffs...

**Comment by gr2m** [comment]
@jeffwilcox if you could share your implementations, that’d be much appreciated. 

I’m thinking about an officially supported plugin architecture for `node-github`, among with official plugins for pagination, caching and throttling. 

I think it would be helpful to have your higher-level library as a great test case that I’d like to accommodate with the new architecture.

**Comment by gr2m** [comment]
Note to self: there is also a `X-Poll-Interval` header which is so far only relevant to the events & notifications APIs: https://developer.github.com/changes/2012-10-26-notifications-api/#polling

**Comment by SGrondin** [comment]
Developer of Bottleneck here. I was thinking about submitting a PR that adds built-in throttling to node-github. By default it would implement Github's official rate limit guidelines and throttle correctly based on the client ID. All of the queueing, concurrency control, etc, mentioned here is already implemented in bottleneck. From my observations, `probot` is not the only project wrapping the github handler and it seems like @bkeepers 's [comment](https://github.com/octokit/node-github/issues/526#issuecomment-289669822) is getting quite a bit of Google traffic too.

Would there be any interest in this?

**Comment by pvdlg** [comment]
If that can help I've implemented a throttling (and retry) feature in [semantic-release/github](https://github.com/semantic-release/github/blob/e1920d3cd76806f4cabaf84a88e6bec68ea2dfe8/lib/get-client.js)

The implementation is based on [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)s to intercept the function calls and executed them with a throttler and retry them if necessary.
The `Proxy` would not be necessary though as it would be implemented directly in the client.

All the calls to any client function goes through a global throttler that limit to one call per second as recommended in [Dealing with abuse rate limits](https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits).

In addition, The GitHub API has different rate limit for the [search](https://developer.github.com/v3/search/#rate-limit) and the [core](https://developer.github.com/v3/#rate-limiting) endpoints. Those limitation are different than the rate abuse recommendation though.
The implementation handle the creation of throttler per limit group. This way the calls to different endpoints belonging to the same rate limit group are throttled together. Each retries are also throttled together based on their rate limit group.
That makes it easy to add a new rate limit group if GitHub adds one in the API.

Each calls are retried on error and are throttled as follow:
- 1 call per second globally
- 30 calls per minutes for the `search` endpoints
- 5000 calls per hour for the `core` endpoints

This implementation doesn't handle `X-RateLimit-Remaining` and `X-RateLimit-Reset`.  The `reservoir ` option of [bottleneck](https://github.com/SGrondin/bottleneck) would be well suited for that. But it might be a bit tricky to handle the situation where the reservoir (set with `X-RateLimit-Remaining` value) would reach 0. Not sure if the client should throw an error or wait for the Rate limit to reset.

**Comment by alextreppass** [comment]
Hi @SGrondin 

> Would there be any interest in this?

Am definitely interested. Currently looking at implementing @bkeepers's snippet above, but it looks like octokit no-longer exposes a `handler` property on the client. 

I wrote a client wrapper class that only exposes the github method calls I need, and used `limiter.schedule` inside those methods instead.

**Comment by gr2m** [comment]
Hey @SGrondin, it’s been a year, but I’m now very interested in implementing the throttling, see also #1099. I’d start out with a plugin so we can move faster and do breaking changes independent of `@octokit/rest` or soon `octokit`. One concern I have is bundle size as this should run nicely in browser’s too. 

If you are still interested I’d love to collaborate on it!

**Comment by SGrondin** [comment]
Wonderful! I'm a big fan of Github integrations and Github tooling and I'd love to be a part of it. Agreed that a plugin would be the way to go. The plugin system looks nice to use. Let's find time to discuss requirements for this plugin, I'll reach out to you on Slack tonight, then we can talk about the next steps / roadmap.

>One concern I have is bundle size as this should run nicely in browser’s too.

This shouldn't be an issue, Bottleneck has zero dependencies and I'd be happy to generate a tiny bundle that skips all clustering logic/lua code. I'm assuming octokit isn't going to be running babel over its dependencies, is the target going to be ES5 despite the bloat it adds?

**Comment by gr2m** [comment]
> I'm assuming octokit isn't going to be running babel over its dependencies, is the target going to be ES5 despite the bloat it adds?

I’ve avoided build steps for node so far, but moving forward I’m not opposed using es6 modules, especially to take advantage of tree shaking. We might also redo everything in Typescript as that lots of GitHubbers use that internally as well as in Probot

**Comment by bkeepers** [comment]
Here's an octokit plugin for throttling that evolved from the snipped above: https://github.com/probot/probot/blob/master/src/github/rate-limiting.ts

**Comment by gr2m** [comment]
closing in favor of #1099 and https://github.com/octokit/plugin-throttling.js
